// @atom-ide-community/nuclide-commons/runtime-info.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRuntimeInformation = getRuntimeInformation;
Object.defineProperty(exports, "__DEV__", {
  enumerable: true,
  get: function () {
    return _env.__DEV__;
  }
});

var _systemInfo = require("./system-info");

var _os = _interopRequireDefault(require("os"));

var _uuid = _interopRequireDefault(require("uuid"));

var _env = require("@atom-ide-community/nuclide-node-transpiler/lib/env");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
let cachedInformation = null;

function getCacheableRuntimeInformation() {
  // eslint-disable-next-line eqeqeq
  if (cachedInformation !== null) {
    return cachedInformation;
  }

  cachedInformation = {
    sessionId: _uuid.default.v4(),
    user: _os.default.userInfo().username,
    osType: (0, _systemInfo.getOsType)(),
    timestamp: 0,
    isClient: !(0, _systemInfo.isRunningInServer)(),
    isDevelopment: _env.__DEV__,
    atomVersion: typeof atom === 'object' ? (0, _systemInfo.getAtomVersion)() : '',
    nuclideVersion: (0, _systemInfo.getNuclideVersion)(),
    installerPackageVersion: 0,
    uptime: 0,
    // TODO (chenshen) fill following information.
    serverVersion: 0
  };
  return cachedInformation;
}

function getRuntimeInformation() {
  const runtimeInformation = { ...getCacheableRuntimeInformation(),
    timestamp: Date.now(),
    uptime: Math.floor(process.uptime() * 1000)
  };
  return runtimeInformation;
}
