// @atom-ide-community/nuclide-commons-ui/Ansi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _anser = _interopRequireDefault(require("anser"));

var _escapeCarriage = _interopRequireDefault(require("escape-carriage"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
function ansiToJSON(input, useClasses) {
  const classes = useClasses == null || !useClasses ? false : useClasses;
  return _anser.default.ansiToJson((0, _escapeCarriage.default)(input), {
    use_classes: classes,
    json: true,
    remove_empty: true
  });
} // make sure


function ansiJSONtoStyleBundle(ansiBundle, colorStyle) {
  const style = {};

  if (ansiBundle.bg) {
    style.backgroundColor = colorStyle != null ? `rgb(${colorStyle[ansiBundle.bg]})` : `rgb(${ansiBundle.bg})`;
  }

  if (ansiBundle.fg) {
    style.color = colorStyle != null ? `rgb(${colorStyle[ansiBundle.fg]})` : `rgb(${ansiBundle.fg})`;
  } else {
    if (colorStyle != null) {
      style.color = `rgb(${colorStyle.default})`;
    }
  }

  return {
    content: ansiBundle.content,
    style
  };
}

function ansiToInlineStyle(text, useClasses, colorStyle) {
  return ansiToJSON(text, useClasses).map(input => ansiJSONtoStyleBundle(input, colorStyle));
}

function defaultRenderSegment({
  key,
  style,
  content
}) {
  return /*#__PURE__*/React.createElement("span", {
    key: key,
    style: style
  }, content);
}

class Ansi extends React.PureComponent {
  render() {
    const {
      useClasses,
      colorStyle,
      children,
      renderSegment = defaultRenderSegment,
      ...passThroughProps
    } = this.props;
    return /*#__PURE__*/React.createElement("code", passThroughProps, ansiToInlineStyle(children, useClasses, colorStyle).map(({
      style,
      content
    }, key) => renderSegment({
      key: String(key),
      style,
      content
    })));
  }

}

exports.default = Ansi;
