// @atom-ide-community/nuclide-commons-ui/BoundSettingsControl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _SettingsControl = _interopRequireDefault(require("./SettingsControl"));

var React = _interopRequireWildcard(require("react"));

var _analytics = require("@atom-ide-community/nuclide-commons/analytics");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class BoundSettingsControl extends React.Component {
  constructor(props) {
    super(props);
    this._observeConfigDisposable = void 0;

    this._onChange = value => {
      (0, _analytics.track)('bound-settings-control-change', {
        keyPath: this.props.keyPath,
        value
      });

      if (this.props.onChangeCallback != null) {
        this.props.onChangeCallback();
      }

      atom.config.set(this.props.keyPath, value);
    };

    this.state = {
      value: atom.config.get(props.keyPath)
    };
  }

  _updateSubscription() {
    if (this._observeConfigDisposable != null) {
      this._observeConfigDisposable.dispose();
    }

    this._observeConfigDisposable = atom.config.onDidChange(this.props.keyPath, ({
      newValue
    }) => {
      this.setState({
        value: newValue
      });
    });
  }

  componentDidMount() {
    this._updateSubscription();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.keyPath !== this.props.keyPath) {
      this.setState({
        value: atom.config.get(this.props.keyPath)
      });

      this._updateSubscription();
    }
  }

  componentWillUnmount() {
    if (this._observeConfigDisposable != null) {
      this._observeConfigDisposable.dispose();
    }
  }

  render() {
    const schema = atom.config.getSchema(this.props.keyPath);
    return /*#__PURE__*/React.createElement(_SettingsControl.default, {
      keyPath: this.props.keyPath,
      title: schema.title,
      value: this.state.value,
      onChange: this._onChange,
      schema: schema,
      hideDetails: this.props.hideDetails
    });
  }

}

exports.default = BoundSettingsControl;
