// @atom-ide-community/nuclide-commons-ui/Dropdown.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DropdownExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Dropdown = require("./Dropdown");

var _SplitButtonDropdown = require("./SplitButtonDropdown");

var _ModalMultiSelect = require("./ModalMultiSelect");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */

/* global alert */
const DropdownExample = (() => {
  const options = [{
    value: 1,
    label: 'One'
  }, {
    value: 2,
    label: 'Two'
  }, {
    value: 3,
    label: 'Three'
  }, {
    value: 4,
    label: 'Four'
  }];
  return () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_Dropdown.Dropdown, {
    options: options,
    value: 2
  }));
})();

const SplitButtonDropdownExample = (() => {
  const options = [{
    value: 1,
    label: 'Build',
    icon: 'tools'
  }, {
    value: 2,
    label: 'Run',
    icon: 'triangle-right',
    selectedLabel: 'Run It!'
  }, {
    value: 3,
    label: 'Rocket',
    icon: 'rocket'
  }, {
    type: 'separator'
  }, {
    value: 4,
    label: 'Squirrel',
    icon: 'squirrel'
  }, {
    value: 5,
    label: 'Beaker',
    icon: 'telescope',
    disabled: true
  }];
  return () => /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_SplitButtonDropdown.SplitButtonDropdown, {
    options: options,
    value: 2,
    onConfirm: // eslint-disable-next-line no-alert
    x => alert(`You selected ${x}!`)
  }));
})();

class ModalMultiSelectExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: [2]
    };
  }

  render() {
    const options = [{
      value: 1,
      label: 'One'
    }, {
      value: 2,
      label: 'Two'
    }, {
      value: 3,
      label: 'Three'
    }, {
      value: 4,
      label: 'Four'
    }];
    return /*#__PURE__*/React.createElement(_ModalMultiSelect.ModalMultiSelect, {
      options: options,
      onChange: value => {
        this.setState({
          value
        });
      },
      value: this.state.value
    });
  }

}

const DropdownExamples = {
  sectionName: 'Dropdowns',
  description: 'For selecting things.',
  examples: [{
    title: 'Dropdown',
    component: DropdownExample
  }, {
    title: 'Split Button Dropdown',
    component: SplitButtonDropdownExample
  }, {
    title: 'Modal Multi-Select',
    component: ModalMultiSelectExample
  }]
};
exports.DropdownExamples = DropdownExamples;
