// @atom-ide-community/nuclide-commons-ui/highlightOnUpdate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.highlightOnUpdate = highlightOnUpdate;

var React = _interopRequireWildcard(require("react"));

var _shallowequal = _interopRequireDefault(require("shallowequal"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */

/**
 * Wraps DecoratedComponent in a special `span` with a configurable classname whenever the
 * component's props change.
 */
function highlightOnUpdate(ComposedComponent,
/**
 * The result of this function determines whether to apply the highlight or not.
 */
arePropsEqual = _shallowequal.default,
/**
 * className used in the wrapper. You can style both `className` and `<className>-highlight`.
 */
className = 'nuclide-ui-highlight-on-render',
/**
 * Delay in ms until the `*-highlight` className gets removed from the wrapper.
 * Effectively throttles the frequency of highlight pulses.
 */
unhighlightDelay = 200) {
  var _temp;

  // $FlowIssue The return type is guaranteed to be the same as the type of ComposedComponent.
  return _temp = class extends React.Component {
    constructor(props) {
      super(props);
      this.showFlash = void 0;
      this.timeout = void 0;
      this.showFlash = false;
    }

    UNSAFE_componentWillUpdate(nextProps, nextState) {
      if (arePropsEqual(nextProps, this.props)) {
        // Skip if prop values didn't actually change.
        return;
      }

      if (this.timeout != null || this.showFlash) {
        // Skip if already scheduled.
        return;
      }

      this.showFlash = true;
      this.timeout = setTimeout(() => {
        this.showFlash = false;
        this.timeout = null;
        this.forceUpdate();
      }, unhighlightDelay);
    }

    render() {
      return /*#__PURE__*/React.createElement("span", {
        className: `${className} ${this.showFlash ? className + '-highlight' : ''}`
      }, /*#__PURE__*/React.createElement(ComposedComponent, this.props));
    }

  }, _temp;
}
