// @atom-ide-community/nuclide-commons-ui/HighlightedCode.example.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HighlightedCodeExamples = void 0;

var React = _interopRequireWildcard(require("react"));

var _Button = require("./Button");

var _HighlightedCode = require("./HighlightedCode");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */

/* global requestAnimationFrame */
class HighlightedCodeExample extends React.Component {
  constructor(...args) {
    super(...args);
    this.state = {
      count: 1
    };

    this._addOneMore = () => {
      // TODO(pelmers): Use react deferred update API when facebook/react/issues/13306 is ready
      requestAnimationFrame(() => {
        // TODO: (wbinnssmith) T30771435 this setState depends on current state
        // and should use an updater function rather than an object
        // eslint-disable-next-line react/no-access-state-in-setstate
        this.setState({
          count: this.state.count + 1
        });
      });
    };
  }

  render() {
    const jsGrammar = atom.grammars.grammarForScopeName('source.js');

    if (jsGrammar == null) {
      return null;
    } // Use our own source code as an example!


    const code = (HighlightedCodeExample.toString() + '\n').repeat(this.state.count); // $FlowIgnore: Not an official API yet.

    const ConcurrentMode = React.unstable_ConcurrentMode;
    return /*#__PURE__*/React.createElement("div", null, "The code below is rendered with async React, so highlighting does not block (no matter how many lines have to be tokenized).", /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement(_Button.Button, {
      onClick: this._addOneMore
    }, "Add more code!"), /*#__PURE__*/React.createElement(ConcurrentMode, null, /*#__PURE__*/React.createElement(_HighlightedCode.HighlightedCode, {
      grammar: jsGrammar,
      code: code,
      style: {
        marginTop: '8px'
      }
    })));
  }

}

const HighlightedCodeExamples = {
  sectionName: 'HighlightedCode',
  description: 'HighlightedCode provides a lighter-weight syntax highlighter for code.\n' + 'It uses the same tokenizer as Atom text editors but ditches the editor.',
  examples: [{
    title: 'HighlightedCode',
    component: HighlightedCodeExample
  }]
};
exports.HighlightedCodeExamples = HighlightedCodeExamples;
