// @atom-ide-community/nuclide-commons-ui/ListView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListView = exports.ListViewItem = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var React = _interopRequireWildcard(require("react"));

var _ignoreTextSelectionEvents = _interopRequireDefault(require("./ignoreTextSelectionEvents"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/**
 * Use ListViewItem in conjunction with ListView.
 */
class ListViewItem extends React.Component {
  _select(value, index, event) {
    if (this.props.onSelect != null) {
      this.props.onSelect(value, index);
    }
  }

  render() {
    const {
      children,
      index,
      value,
      ...remainingProps
    } = this.props;
    return /*#__PURE__*/React.createElement("div", _extends({
      className: "nuclide-ui-listview-item"
    }, remainingProps, {
      onClick: (0, _ignoreTextSelectionEvents.default)(this._select.bind(this, value, index))
    }), children);
  }

}

exports.ListViewItem = ListViewItem;

class ListView extends React.Component {
  constructor(...args) {
    super(...args);

    this._handleSelect = (value, index, event) => {
      if (this.props.selectable && this.props.onSelect != null) {
        this.props.onSelect(index, value);
      }
    };
  }

  render() {
    const {
      children,
      alternateBackground,
      selectable
    } = this.props;
    const renderedItems = React.Children.map(children, (child, index) => /*#__PURE__*/React.cloneElement(child, {
      index,
      onSelect: this._handleSelect
    }));
    const className = (0, _classnames.default)({
      'native-key-bindings': true,
      'nuclide-ui-listview': true,
      'nuclide-ui-listview-highlight-odd': alternateBackground,
      'nuclide-ui-listview-selectable': selectable
    });
    return /*#__PURE__*/React.createElement("div", {
      className: className,
      tabIndex: -1
    }, renderedItems);
  }

}

exports.ListView = ListView;
