// @atom-ide-community/nuclide-commons-ui/PulseButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
class PulseButton extends React.Component {
  render() {
    const {
      ariaLabel,
      className,
      isSelected,
      size = 10,
      style,
      onClick,
      onMouseOver,
      onMouseLeave
    } = this.props;
    return /*#__PURE__*/React.createElement("a", {
      className: (0, _classnames.default)({
        'nuclide-ui-pulse-button': true,
        selected: isSelected
      }, className),
      style: {
        height: size,
        width: size,
        ...style
      },
      href: "#",
      "aria-label": ariaLabel,
      onClick: onClick,
      onMouseOver: onMouseOver,
      onMouseLeave: onMouseLeave,
      role: "button"
    }, /*#__PURE__*/React.createElement("div", {
      className: "nuclide-ui-pulse-button--inner-circle"
    }), /*#__PURE__*/React.createElement("div", {
      className: "nuclide-ui-pulse-button--cover-circle"
    }));
  }

}

exports.default = PulseButton;
