// @atom-ide-community/nuclide-commons-ui/Tooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tooltip = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */
// Our custom ref component invokes a callback after updating
// if the child component has changed. We need a custom ref component
// because stateless functional components can't have refs
class RefWrapper extends React.Component {
  constructor(...args) {
    super(...args);
    this._el = null;
  }

  componentDidMount() {
    this._updateElement();
  }

  componentWillUnmount() {
    if (this._el != null) {
      this._el = null;
      this.props.customRef(null);
    }
  }

  _updateElement() {
    const el = _reactDom.default.findDOMNode(this);

    if (el !== this._el) {
      this._el = el;
      this.props.customRef(null);
      this.props.customRef(el);
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.children !== prevProps.children) {
      this._updateElement();
    }
  }

  render() {
    return React.Children.only(this.props.children);
  }

}

class Tooltip extends React.Component {
  constructor(props) {
    super(props);
    this._tooltip = void 0;
    this._element = void 0;

    this._displayTooltip = element => {
      if (this._tooltip != null) {
        this._tooltip.dispose();
      }

      if (element != null) {
        this._element = element; // $FlowFixMe - HTMLElement is incompatible with Element

        this._tooltip = atom.tooltips.add(element, {
          title: this.props.title || '',
          delay: this.props.delay
        });
      }
    };

    this._element = null;
  }

  componentDidUpdate(prevProps) {
    if (this._element != null && this.props.title !== prevProps.title) {
      this._displayTooltip(this._element);
    }
  }

  render() {
    return /*#__PURE__*/React.createElement(RefWrapper, {
      customRef: this._displayTooltip
    }, this.props.children);
  }

}

exports.Tooltip = Tooltip;
Tooltip.defaultProps = {
  delay: 0
};
