// atom-ide-base/commons-ui/float-pane/MarkdownView.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderMarkdown = exports.MarkdownView = void 0;
const React = __importStar(require("react"));
const dompurify_1 = __importDefault(require("dompurify"));
const MarkdownRenderer_1 = require("../MarkdownRenderer");
/**
 * A react component that can hosts markdown texts
 */
class MarkdownView extends React.Component {
    constructor() {
        super(...arguments);
        this.state = { markdown: "" };
    }
    render() {
        return (React.createElement("div", { className: this.props.containerClassName, onWheel: (e) => this.onMouseWheel(e) },
            React.createElement("div", { className: this.props.contentClassName, dangerouslySetInnerHTML: {
                    __html: this.state.markdown,
                } })));
    }
    /**
     * handles the mouse wheel event to enable scrolling over long text
     * @param evt the mouse wheel event being triggered
     */
    onMouseWheel(evt) {
        evt.stopPropagation();
    }
    /**
      Calls `getDocumentationHtml` to convert Markdown to markdown
    */
    async componentDidMount() {
        var _a;
        this.setState({
            markdown: (_a = (await renderMarkdown(this.props.markdown, this.props.grammarName, this.props.renderer))) !== null && _a !== void 0 ? _a : "",
        });
    }
}
exports.MarkdownView = MarkdownView;
/**
 * convert the markdown documentation to markdown
 * @param markdownTexts the documentation text in markdown format to be converted
 * @param grammarName  the default grammar used for embedded code samples
 * @param renderer markdown service to be used for rendering
 * @return a promise object to track the asynchronous operation
 */
async function renderMarkdown(markdownTexts, grammarName, renderer) {
    var _a, _b;
    if (grammarName === void 0) { grammarName = ((_b = (_a = atom.workspace.getActiveTextEditor()) === null || _a === void 0 ? void 0 : _a.getGrammar().scopeName) === null || _b === void 0 ? void 0 : _b.toLowerCase()) || ""; }
    if (markdownTexts === undefined) {
        return null;
    }
    let markdownText = "";
    // if Array
    if (Array.isArray(markdownTexts)) {
        if (markdownTexts.length === 0) {
            return null;
        }
        markdownText = markdownTexts.join("\r\n");
    }
    // if string
    else {
        markdownText = markdownTexts;
    }
    if (renderer) {
        return dompurify_1.default.sanitize(await renderer.render(markdownText, grammarName));
    }
    else {
        // Use built-in markdown renderer (it already does sanitization)
        const render = await MarkdownRenderer_1.getMarkdownRenderer();
        return render(markdownText, grammarName);
    }
}
exports.renderMarkdown = renderMarkdown;
