// atom-ide-base/commons-ui/float-pane/selectable-overlay.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.copyListener = exports.makeOverLayCopyable = exports.blurEditor = exports.focusEditor = exports.overlayFocusFix = exports.makeOverlaySelectable = void 0;
/** makes the text selectable and copyable
 *
 * Note: you can directly add `user-select: text` (and `pointer-events: all`) in CSS for better performance
 */
function makeOverlaySelectable(editor, overlayElement, focusFix = true) {
    // allow the browser to handle selecting
    overlayElement.setAttribute("tabindex", "-1");
    // make it selectable
    if (!overlayElement.style.userSelect || overlayElement.style.userSelect === "none") {
        overlayElement.style.userSelect = "text";
    }
    if (focusFix) {
        // fix overlay focus issue
        overlayFocusFix(editor, overlayElement);
    }
    // add copy keybindings
    overlayElement.classList.add("native-key-bindings");
}
exports.makeOverlaySelectable = makeOverlaySelectable;
/**
 * - focus on the datatip once the text is selected (cursor gets disabled temporarily)
 * - remove focus once mouse leaves
 */
function overlayFocusFix(editor, element) {
    const editorComponent = atom.views.getView(editor).getComponent();
    element.addEventListener("mousedown", () => {
        blurEditor(editorComponent);
        element.addEventListener("mouseleave", () => {
            focusEditor(editorComponent);
        });
    });
}
exports.overlayFocusFix = overlayFocusFix;
function focusEditor(editorComponent) {
    // @ts-ignore internal api
    editorComponent === null || editorComponent === void 0 ? void 0 : editorComponent.didFocus();
}
exports.focusEditor = focusEditor;
function blurEditor(editorComponent) {
    // @ts-ignore internal api
    editorComponent === null || editorComponent === void 0 ? void 0 : editorComponent.didBlurHiddenInput({
        relatedTarget: null,
    });
}
exports.blurEditor = blurEditor;
/*
██████  ███████ ██████  ██████  ███████  ██████  █████  ████████ ███████ ██████
██   ██ ██      ██   ██ ██   ██ ██      ██      ██   ██    ██    ██      ██   ██
██   ██ █████   ██████  ██████  █████   ██      ███████    ██    █████   ██   ██
██   ██ ██      ██      ██   ██ ██      ██      ██   ██    ██    ██      ██   ██
██████  ███████ ██      ██   ██ ███████  ██████ ██   ██    ██    ███████ ██████
*/
/** @deprecated use `makeOverlaySelectable` instead.
 *
 * Makes the overlay component copyable
 * - you should call `makeOverlaySelectable` before this
 * - If your element already has mouseenter and mouseleav listeners, directly use `copyListener`
 */
function makeOverLayCopyable(element) {
    element.addEventListener("mouseenter", () => {
        element.addEventListener("keydown", copyListener);
    });
    element.addEventListener("mouseleave", () => {
        element.removeEventListener("keydown", copyListener);
    });
}
exports.makeOverLayCopyable = makeOverLayCopyable;
/** @deprecated use `makeOverlaySelectable` instead.
 *
 * A manual copy listener
 * Usage. Add the listener to your mouse enter and mouseleave listeners
   ```ts
   element.addEventListener("mouseenter", () => {element.addEventListener("keydown", copyListener)}`
   element.addEventListener("mouseleave", () => {element.removeEventListener("keydown", copyListener)}`
   ```
*/
async function copyListener(event) {
    var _a, _b;
    event.preventDefault();
    if (event.ctrlKey && event.key === "c") {
        const text = (_b = (_a = document.getSelection()) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : "";
        await navigator.clipboard.writeText(text);
    }
} // TODO we should not need to manually listen for copy paste
exports.copyListener = copyListener;
