// atom-ide-base/commons-ui/float-pane/SnippetView.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSnippetHtml = exports.SnippetView = void 0;
const React = __importStar(require("react"));
const dompurify_1 = __importDefault(require("dompurify"));
const MarkdownRenderer_1 = require("../MarkdownRenderer");
/**
 * A React component that hosts a code snippet with syntax highlighting
 */
class SnippetView extends React.Component {
    constructor() {
        super(...arguments);
        this.state = { snippet: "" };
    }
    render() {
        return (React.createElement("div", { className: this.props.containerClassName, onWheel: (e) => this.onMouseWheel(e) },
            React.createElement("div", { className: this.props.contentClassName, dangerouslySetInnerHTML: {
                    __html: this.state.snippet,
                } })));
    }
    /**
     * handles the mouse wheel event to enable scrolling over long text
     * @param evt the mouse wheel event being triggered
     */
    onMouseWheel(evt) {
        evt.stopPropagation();
    }
    async componentDidMount() {
        var _a;
        this.setState({
            snippet: (_a = (await getSnippetHtml(this.props.snippet, this.props.grammarName, this.props.renderer))) !== null && _a !== void 0 ? _a : "",
        });
    }
}
exports.SnippetView = SnippetView;
const regexPremeable = /^\s*<([!?])([a-z]+)?\s*/i;
const regexLSPPrefix = /^\((method|property|parameter|alias)\)\W/;
/**
 * converts a given code snippet into syntax formatted HTML
 * @param snippets the code snippet to be converted
 * @param grammarName the name of the grammar to be used for syntax highlighting
 * @param renderer markdown service to be used for rendering
 * @return a promise object to track the asynchronous operation
 */
async function getSnippetHtml(snipetsGiven, grammarName, renderer) {
    var _a, _b;
    if (grammarName === void 0) { grammarName = ((_b = (_a = atom.workspace.getActiveTextEditor()) === null || _a === void 0 ? void 0 : _a.getGrammar().scopeName) === null || _b === void 0 ? void 0 : _b.toLowerCase()) || ""; }
    if (snipetsGiven === undefined) {
        return null;
    }
    let snippets = snipetsGiven;
    // if string
    if (typeof snippets === "string") {
        snippets = [snippets];
    }
    // if Array
    if (Array.isArray(snippets)) {
        if (snippets.length === 0) {
            return null;
        }
        const markdown = snippets
            .map((snippet) => {
            const snp = snippet
                .replace(regexPremeable, "") // remove any preamble from the line
                .replace(regexLSPPrefix, ""); // remove LSP prefix
            return `\`\`\`\n${snp}\n\`\`\``;
        })
            .join("\n");
        if (renderer) {
            return dompurify_1.default.sanitize(await renderer.render(markdown, grammarName));
        }
        else {
            // Use built-in markdown renderer (it already does sanitization)
            const render = await MarkdownRenderer_1.getMarkdownRenderer();
            return render(markdown, grammarName);
        }
    }
    else {
        return null;
    }
}
exports.getSnippetHtml = getSnippetHtml;
